plugin/src/main/java/org/opensearch/ml/action/agents/TransportRegisterAgentAction.java

```diff
@@ -60,10 +66,42 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<MLRegi
 
     private void registerAgent(MLAgent agent, ActionListener<MLRegisterAgentResponse> listener) {
         Instant now = Instant.now();
-        MLAgent mlAgent = agent.toBuilder().createdTime(now).lastUpdateTime(now).build();
+        String llmId;
+        if (agent.getType().equalsIgnoreCase(MLAgentType.CONVERSATIONAL.name()) && agent.getLlm() != null) {
+            llmId = agent.getLlm().getModelId();
+        } else {
+            llmId = null;
+        }
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            if (llmId != null) {
+                String[] excludes = new String[] { MLModel.MODEL_CONTENT_FIELD, MLModel.OLD_MODEL_CONTENT_FIELD };
+                mlModelManager.getModel(llmId, null, excludes, ActionListener.wrap(mlModel -> {
+                    processModelAndRegisterAgent(agent, mlModel.getIsHidden(), now, listener, context);
+                }, e -> {
+                    log.error("Failed to retrieve the ML model with ID: " + llmId, e);
+                    listener.onFailure(e);
+                }));
+            } else {
+                processModelAndRegisterAgent(agent, false, now, listener, context);
+            }
+
+        } catch (Exception e) {
+            log.error("Failed to get the model information " + llmId, e);
+            listener.onFailure(e);
+        }
+    }
+
+    private void processModelAndRegisterAgent(
+        MLAgent agent,
+        Boolean isHiddenModel,
+        Instant now,
+        ActionListener<MLRegisterAgentResponse> listener,
+        ThreadContext.StoredContext context
+    ) {
+        MLAgent mlAgent = agent.toBuilder().createdTime(now).lastUpdateTime(now).isHidden(isHiddenModel).build();
         mlIndicesHandler.initMLAgentIndex(ActionListener.wrap(result -> {
             if (result) {
-                try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+                try (context) { // Using try-with-resources to ensure 'context' is restored
```

Comment:
Can we not pass the context along into nested code, that makes hard to understand what's the impact scope of context

Reply:
I'm passing the context as a method parameter. That was I can re-use the code.

Reply:
Not quite understand, which code you are re-using? 

We should minimize the scope of `stashContext`.  `stashContext` in Line 75 is for get model in line 78, the action listener in line 78 should use runAfter to restore context to minimize the scope of line 75. 

In this action listener of `initMLAgentIndex` , we should stash it again so it's clear the line 75 is for get model, here we stash again for line 109.

Reply:
done